docs/proposals/api-types-package-structure.md

```diff
@@ -0,0 +1,381 @@
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a types.go file contains an "internal" version of the group's
+objects. This file is used as a destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
+little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers the v1 version of this group, its
+conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/install/conversion`
+
+This package contains conversion functions, and if imported, will register them.
+
+There are two sorts of conversion functions: those which could be automatically
+generated, and those which have manual additions. All the conversion functions
+convert to & from the types in the parent `apis/groups` and `apis/groups/v1`
+directories.
+
+### Package: `apis/group/v1/install/defaulting`
+
+This package contains defaulting functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+There are two sorts of defaulting functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for `// +default=<value>` comments
+preceding the types or their embedded struct members. Similarly, a
+`// +default=custom` marker will force a stub to be generated. Stubs will be
+generated with a `panic("write me!")` marker to ensure that they are populated.
+
+### Package: `apis/group/v1/install/validation`
+
+This package contains validation functions, and if imported, will register them.
+
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+The allowed imports are constrained; in particular, nothing that would let you
+do RPCs to look up other cluster state is allowed. Validation functions validate
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere. (TODO: copy-pasta this 2x above)
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validate_min=<numeric value>,validate_max=<numeric value>`
+* `// +validate_regexp=<regexp>`
+* `// +validate_exact=<exact value>`
```

Comment:
Yeah, this is not a specific proposal, it's more of an example of the sort of thing I want us to do.

I do *not* want the auto-generated-validation to be turing complete, I don't want anything complicated enough that someone has to look up the documentation to understand what it will do. But we should debate this in a narrower document.